lib/ansible/module_utils/ec2.py

```diff
@@ -554,6 +554,12 @@ def _hashable_policy(policy, policy_list):
          ('Version', (u'2012-10-17',)))]
 
     """
+    # Amazon will automatically convert bool and int to strings for us
+    if isinstance(policy, bool):
```

Comment:
In general this looks good to me. `compare_policies` is used in quite a few modules for comparing general dictionaries so it's possible one would want to actually compare booleans/ints/strings. I wonder if the guts of this function should be refactored so there's a more generally-safe function without the more policy-specific anomalies.

Reply:
@s-hertel:
The useful behaviour here is that it
- treats arrays/lists as an unordered set rather than an ordered list.
- makes a list with one entry equal to just the single entry.

That feels like risky behaviour to rely for anything other than an IAM policy, but I wouldn't be surprised if someone had done so.

How about making this behaviour optional with a flag (convert_to_strings ?) which defaults to true, so that when you're using it for the 'obvious' behaviour it adheres to the rules defined by Amazon for IAM policies, but if you wanted something slightly different, it is still there.

Reply:
Rather than leaving this just as an IRC conversation.

aws_s3_cors and aws_waf_condition are the two examples I can find.  Only aws_waf_condition has a test suite ( aws_waf_web_acl )

Neither of which support raw booleans (as far as I can see), and I can't see any examples where quoting raw numbers should cause a problem...

Reply:
Looking at and testing both (manually in the case of aws_s3_cors), I think we're good.

Reply:
@jillr aws_waf_web_acl and aws_codepipeline are two I have seen do this. I believe the ELB modules were also considering using it. It's a convenient way to compare dictionaries since we can't use OrderedDict yet.

@tremble But it's not just about modules in Ansible, you know, because any third party module can be relying on module_utils. That's why we can't rename or move functions unless we leave the old function to call the new function for backward-compatibility. That said, this is pretty minimal and aws_waf_web_acl looks broken for other reasons, and it's better to get changes like this in earlier in the cycle than later so people can report issues.